File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/repository/PostRepository.kt
package com.bubbler.android.data.repository

import com.bubbler.android.core.auth.TokenStore
import com.bubbler.android.core.network.ApiClient
import com.bubbler.android.core.network.Endpoints
import com.bubbler.android.data.model.Post
import kotlinx.serialization.Serializable
import okhttp3.RequestBody.Companion.toRequestBody

/** Create / update / delete posts and topic mutations. */
open class PostRepository(
    private val apiClient: ApiClient,
    private val tokenStore: TokenStore,
) {
    open suspend fun createPost(content: String, topic: String): Post {
        val payload = apiClient.json.encodeToString(
            CreatePostBody.serializer(),
            CreatePostBody(post = content, topic = topic),
        )
        return apiClient.request(
            path = Endpoints.USER_ME_POSTS,
            method = "POST",
            token = tokenStore.requireAccessToken(),
            body = payload.toRequestBody(ApiClient.JSON_MEDIA_TYPE),
            contentType = "application/json",
        )
    }

    open suspend fun updatePost(id: String, content: String) {
        val payload = apiClient.json.encodeToString(
            UpdatePostBody.serializer(),
            UpdatePostBody(post = content),
        )
        apiClient.executeIgnoringBody(
            path = Endpoints.userPost(id),
            method = "PUT",
            token = tokenStore.requireAccessToken(),
            body = payload.toRequestBody(ApiClient.JSON_MEDIA_TYPE),
            contentType = "application/json",
        )
    }

    open suspend fun deletePost(id: String) {
        apiClient.executeIgnoringBody(
            path = Endpoints.userPost(id),
            method = "DELETE",
            token = tokenStore.requireAccessToken(),
        )
    }

    open suspend fun addPostTopic(postId: String, topic: String): Post {
        val payload = apiClient.json.encodeToString(
            PostTopicMutationBody.serializer(),
            PostTopicMutationBody(topic = topic),
        )
        return apiClient.request(
            path = Endpoints.userPostTopics(postId),
            method = "POST",
            token = tokenStore.requireAccessToken(),
            body = payload.toRequestBody(ApiClient.JSON_MEDIA_TYPE),
            contentType = "application/json",
        )
    }

    open suspend fun removePostTopic(postId: String, topic: String): Post =
        apiClient.request(
            path = Endpoints.userPostTopic(postId, topic),
            method = "DELETE",
            token = tokenStore.requireAccessToken(),
        )
}

@Serializable
private data class CreatePostBody(
    val post: String,
    val topic: String,
)

@Serializable
private data class UpdatePostBody(
    val post: String,
)

@Serializable
private data class PostTopicMutationBody(
    val topic: String,
)

File: bubbler_app/lib/data/repositories/post_repository.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_exception.dart';
import '../../core/api/endpoints.dart';
import '../models/post.dart';

/// Create / update / delete posts and topic mutations.
class PostRepository {
  PostRepository(this._client);

  final ApiClient _client;

  /// Creates a post for the signed-in user (`POST /user/me/posts`).
  Future<Post> createPost({
    required String content,
    required String topic,
  }) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMePosts,
      method: 'POST',
      data: {
        'post': content,
        'topic': topic,
      },
      contentType: 'application/json',
    );
    return Post.fromJson(_asJsonMap(data));
  }

  /// Updates post body (`PUT /user/me/posts/{id}`).
  Future<void> updatePost({
    required String id,
    required String content,
  }) async {
    await _client.authorizedRequest(
      path: Endpoints.userMePost(id),
      method: 'PUT',
      data: {'post': content},
      contentType: 'application/json',
    );
  }

  /// Deletes a post (`DELETE /user/me/posts/{id}`).
  Future<void> deletePost(String id) async {
    await _client.authorizedRequest(
      path: Endpoints.userMePost(id),
      method: 'DELETE',
    );
  }

  /// Adds a topic tag (`POST /user/me/posts/{id}/topics`).
  Future<Post> addPostTopic({
    required String postId,
    required String topic,
  }) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMePostTopics(postId),
      method: 'POST',
      data: {'topic': topic},
      contentType: 'application/json',
    );
    return Post.fromJson(_asJsonMap(data));
  }

  /// Removes a topic tag (`DELETE /user/me/posts/{id}/topics/{topic}`).
  Future<Post> removePostTopic({
    required String postId,
    required String topic,
  }) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMePostTopic(postId, topic),
      method: 'DELETE',
    );
    return Post.fromJson(_asJsonMap(data));
  }

  static Map<String, dynamic> _asJsonMap(dynamic data) {
    if (data is Map<String, dynamic>) {
      return data;
    }
    if (data is Map) {
      return Map<String, dynamic>.from(data);
    }
    throw const ApiInvalidResponse();
  }
}
